pkg/aggregator/mutating_walker.go

```diff
@@ -0,0 +1,513 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aggregator
+
+import (
+	_ "net/http/pprof"
+
+	"github.com/go-openapi/spec"
+)
+
+// Run a walkRefCallback method on all references of an OpenAPI spec, replacing the values.
+type mutatingReferenceWalker struct {
+	// walkRefCallback will be called on each reference. Do not mutate the input, always create a copy first and return that.
+	walkRefCallback func(ref *spec.Ref) *spec.Ref
+}
+
+// replaceReferences rewrites the references without mutating the input.
+// The output might share data with the input.
+func replaceReferences(walkRef func(ref *spec.Ref) *spec.Ref, sp *spec.Swagger) *spec.Swagger {
+	walker := &mutatingReferenceWalker{walkRefCallback: walkRef}
+	return walker.Start(sp)
+}
+
+func (w *mutatingReferenceWalker) walkSchema(schema *spec.Schema) *spec.Schema {
+	if schema == nil {
+		return nil
+	}
+
+	orig := schema
+	clone := func() {
```

Comment:
can you share the code between mutating and non-mutating walkers by defining this function at the type level?

Reply:
Then obviously have one type with two different factory function.

Reply:
I see, the difference is not this clone function but is the `mutatingReferenceWalker`. I think there is still a chance to share most of the code by defining a new type that implements both mutatingReferenceWalker using a readonlyReferenceWalker.

Reply:
We could implement the readonly version on-top of the mutating one. But I don't see much sense the other way around.

Reply:
That's what I meant, readonly one be on top of mutating one.